mlflow/langchain/__init__.py

```diff
@@ -0,0 +1,385 @@
+                        By default, the function waits for five minutes.
+                        Specify 0 or None to skip waiting.
+    :param pip_requirements: {{ pip_requirements }}
+    :param extra_pip_requirements: {{ extra_pip_requirements }}
+    :param metadata: Custom metadata dictionary passed to the model and stored in the MLmodel file.
+
+                     .. Note:: Experimental: This parameter may change or be removed in a future
+                                             release without warning.
+    :return: A :py:class:`ModelInfo <mlflow.models.model.ModelInfo>` instance that contains the
+             metadata of the logged model.
+    """
+    import langchain
+
+    if type(lc_model) != langchain.chains.llm.LLMChain:
+        raise TypeError(
+            "MLflow langchain flavor only supports logging langchain.chains.llm.LLMChain "
+            + f"instances, found {type(lc_model)}"
+        )
+    _SUPPORTED_LLMS = {langchain.llms.openai.OpenAI, langchain.llms.huggingface_hub.HuggingFaceHub}
+    if type(lc_model.llm) not in _SUPPORTED_LLMS:
+        logger.warning(
+            "MLflow does not guarantee support for LLMChains outside of HuggingFaceHub and "
+            "OpenAI, found %s",
+            str(type(lc_model.llm)),
+        )
+    return Model.log(
+        artifact_path=artifact_path,
+        flavor=mlflow.langchain,
+        registered_model_name=registered_model_name,
+        lc_model=lc_model,
+        conda_env=conda_env,
+        code_paths=code_paths,
+        signature=signature,
+        input_example=input_example,
+        await_registration_for=await_registration_for,
+        pip_requirements=pip_requirements,
+        extra_pip_requirements=extra_pip_requirements,
+        metadata=metadata,
+    )
+
+
+def _save_model(model, path):
+    model.save(path)
+
+
+def _load_model(path):
+    from langchain.chains.loading import load_chain
+
+    model = load_chain(path)
+    return model
+
+
+class _LangChainModelWrapper:
+    def __init__(self, lc_model):
+        self.lc_model = lc_model
+
+    def predict(self, data: List[Union[str, Dict[str, str], pd.DataFrame]]) -> List[str]:
+        if isinstance(data, pd.DataFrame):
+            messages = data.to_dict(orient="records")
+        elif isinstance(data, list) and all(isinstance(d, dict) for d in data):
+            messages = data
+        else:
+            raise mlflow.MlflowException.invalid_parameter_value(
+                "Input must be a pandas DataFrame or a list of dictionaries",
+            )
+        return [self.lc_model.run(messages)]
+
+
+class _TestLangChainWrapper(_LangChainModelWrapper):
+    """
+    A wrapper class that should be used for testing purposes only.
+    """
+
+    def predict(self, data):
+        from tests.langchain.test_langchain_model_export import _mock_async_request
+
+        with _mock_async_request():
+            return super().predict(data)
+
+
+def _load_pyfunc(path):
```

Comment:
`_load_pyfunc` seems unused. Delete and remove the env var `_MLFLOW_OPENAI_TESTING`

Reply:
Its a required function that the pyfunc calls to load the model. Its not called internally in the langchain code.